Show/hide toolbar and handles based on device kind

mdebbar opened the pull request:
## Description

When the user is interacting with a text field through a mouse, we don't want to show the toolbar nor the selection handles. They should only be shown when the user is interacting through a touch screen.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [ ] No, this is *not* a breaking change.

Fixes https://github.com/flutter/flutter/issues/31629

## TODO
- [x] Implement for material text field.
  - [x] Write tests.
- [x] Implement for cupertino text field.
  - [x] Write tests.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

mdebbar committed:
Show/hide toolbar and handles based on device kind

xster commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1099,6 +1099,13 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     }
   }
 
+  bool _isSelectionOverlayEnabled = true;
```
This is just e a editabletext. It shouldn't know so much about whether to show the toolbar based on the platform and pointer type etc etc. Just let its parent call or not call showToolbar() based on the appropriate platform+device information.

mdebbar replied:
Agreed. It seems that the logic is still spread across editable text and text field now. I did a clean up and put all the show/hide logic in the text field. PTAL.

xster commented:
I don't have a good idea of how useful knowing the kind will be in general for all gestures. cc @dkwingsmt, @gspencergoog. But if it's all very text fields / TextSelectionGestureDetector specific, you can just add another Listener to TextSelectionGestureDetector that specifically check for pointer kinds coming in and out and pass that back out via a separate callback. For instance, you can pass a pointers tracker into TextSelectionGestureDetector while constructing, and on callbacks of the other callbacks, its parent can ask the tracker it passed in if it's all exclusively tap or mouse and do something special if it's a mix (though it'd be weird that you get a callback at all in that case).

mdebbar commented:
@xster I like the idea of a having a tracker for device kind. I just updated the PR with the changes you recommended. Also added a bunch of tests to make sure the behavior toolbar/handles doesn't break.

mdebbar committed:
Same for Cupertino text field

xster commented on packages/flutter/lib/src/widgets/gesture_detector.dart:
```diff
@@ -580,6 +580,12 @@ class GestureDetector extends StatelessWidget {
   }
 }
 
+/// Keeps track of the last [PointerDeviceKind] that initiated a gesture.
```
I think it's a slightly 'misleading' usage of ValueNotifier since there are no usages of the Listenable interface or listeners etc. Do you intend to expand its usage in the future?

Also, since we were saying we're going this route to avoid having to design a whole generic API around tracking the kind type (since we think it's only useful in the text selection domain), these knowledge shouldn't leak outside of the text selection classes. 

I would just create a listener that composes a Listener above the TextSelectionGestureDetector that doesn't participate in the arena or gestures etc and that only keeps track of the last primary pointer's kind and reports back to the text fields that injected the tracker to the TextSelectionGestureDetector.

mdebbar replied:
Ahh, I misunderstood your previous comment :P
I switched now to using `Listener` and it's much cleaner indeed. I'll push it right away.

mdebbar committed:
use Listener() to track device kind

xster commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -773,20 +811,25 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
-              child: _addTextDependentAttachments(paddedEditable, textStyle, placeholderStyle),
+            child: Listener(
+              onPointerDown: (PointerDownEvent event) {
+                _lastUsedDeviceKind = event.kind;
```
You should decide on this, though while we want to really keep this simple, this may be a bit too simple :D. I think you might still want a little bit of the pointer tracking here. Unless you think the pointer events will never be in a situation where the kinds will mix.

mdebbar replied:
This is by design. We only want to take into account the actions that "trigger" the text field. All actions that "trigger" the text field start with a PointerDown event (tap, long press, drag, long-press-drag, etc).

See the [internal Q/A](https://yaqs.googleplex.com/eng/q/4519774183489536) for the opinion of the Material team on this.

xster replied:
This is really an edge case, but I was asking what is supposed to happen if you held down and dragged a finger on Android, selecting some words, and right before you lift the finger, there was also a mouse click somewhere. 

By your rule above, the 'trigger' is the touch, therefore it should show the toolbar, but by the current implementation, the last kind is mouse and there won't be a toolbar.

xster commented on packages/flutter/test/cupertino/text_field_test.dart:
```diff
@@ -1948,6 +1948,193 @@ void main() {
+    // Long press to trigger the text field.
+    await tester.longPress(find.byType(CupertinoTextField));
+    await tester.pump();
+
```
We should also check that long press always produces collapsed selection (thus no handles)

xster reviewed (commented):
This is much better now. Thanks!

mdebbar commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -541,12 +553,35 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
+      _editableText.showToolbar();
+  }
+
+  bool _shouldShowHandles(SelectionChangedCause cause) {
```
There are cases where things bubble in then out, but many selection changes originate from render editable (e.g. using a keyboard on Android, or semantics).

mdebbar commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -541,12 +553,35 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
+
+  bool _shouldShowHandles(SelectionChangedCause cause) {
+    // When the text field was triggered by anything other than a touch screen,
+    // we shouldn't show the selection handles.
```
You're right. A pen (stylus) should also trigger the selection overlay because it's still interacts through the touch screen. Do you think `invertedStylus` should also trigger the selection overlay? I don't have a pen at the moment to try this.

mdebbar commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -541,12 +553,35 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
+      return false;
+    if (cause == SelectionChangedCause.keyboard)
+      return false;
+
```
The reason I formatted them this way is because the first two return false while the 3rd and 4th return true. I'll add an empty line between all of them.

mdebbar committed:
Address comments

xster commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -541,12 +556,36 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
+    if (cause == SelectionChangedCause.keyboard)
+      return false;
+
+    if (cause == SelectionChangedCause.longPress)
```
This might need a comment above. Isn't long press selections on iOS always collapsed?

xster commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -573,7 +612,10 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
+      _editableText?.bringIntoView(selection.base);
+    }
+    if (_shouldShowHandles(cause)) {
+      _editableText.showHandles();
```
Is this an API breaking change? i.e. if I previously used EditableText as is, it won't show handles anymore after I flutter upgrade?

xster commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -387,10 +396,17 @@ class TextSelectionOverlay {
+    }
+  }
 
+  /// Hides the toolbar.
```
Add a bit more doc here and with `hide` to describe their differences

xster reviewed (approved):
LGTM

dnfield commented:
@mdebbar will you be addressing this feedback?

gspencergoog commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -773,20 +819,23 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with AutomaticK
-              behavior: HitTestBehavior.translucent,
-              child: _addTextDependentAttachments(paddedEditable, textStyle, placeholderStyle),
+            child: Listener(
+              onPointerDown: _updateLastUsedDeviceKind,
```
It seems like to @dkwingsmt and me that a better way to do this would be to add the pointer kind to the details for the gesture detector callbacks, and look at the device kind when needed.  Then you wouldn't need the extra listener, or to maintain a cache of the device kind.

xster replied:
I gave him the opposite advice previously :D

https://github.com/flutter/flutter/pull/29683#issuecomment-475003862

because I wasn't sure where you guys wanted to go with altering signatures everywhere. If you guys were going there anyway, then what you're suggesting makes sense.

gspencergoog replied:
Oh, sorry, I didn't notice that.  Yeah, we're going to have to go there anyhow, since there are going to be other cases of needing to distinguish.  The good thing is that it won't alter a lot of the signatures if there are details objects already, since we can just add it to the details that are already provided.

xster replied:
Sounds reasonable. Sorry for the back and forth @mdebbar.

mdebbar replied:
@gspencergoog thanks for chiming in. @xster it's all good :)

I'll have the necessary changes ready for review by EOD.

mdebbar committed:
Change approach

mdebbar commented:
@xster @gspencergoog I updated the PR. Some tests are still failing but I would like to get a review while I fix those tests. From a quick glance, they seem to be newly added tests for `EditableText` showing the handles. Since the logic for showing/hiding the handles is now in `TextField`, I'll port those tests over.

mdebbar commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -412,7 +427,7 @@ abstract class PrimaryPointerGestureRecognizer extends OneSequenceGestureRecogni
       initialPosition = event.position;
       if (deadline != null)
-        _timer = Timer(deadline, didExceedDeadline);
+        _timer = Timer(deadline, () => didExceedDeadlineWithEvent(event));
```
This isn't ideal. If you have better ideas on how to do it, I'm all ears :)

xster replied:
Ah, this is unfortunate. Though this event isn't entirely correct either right? Since you're sending something from the past and the past event also has a timeStamp. 

I haven't validated it but I suspect each onLongPressDown is preceded by a onTapDown or something. From the TextSelectionGestureDetector's perspective, it may be enough?

mdebbar replied:
Unfortunately, this is how the tap recognizer works. It doesn't send the tap down event until either a tap up occurs, or the deadline is exceeded. And in that case, the tap down is always from the past (even before my change). All my change does is it adds the `kind` field to the event.

mdebbar committed:
Fix tests

xster reviewed (commented):
Thanks for changing this around. Sorry again for the back and forth.

goderbauer commented:
@mdebbar Are you still planning to get back to this? The PR is getting kinda old...

mdebbar committed:
Address comments

mdebbar committed:
Fix tests
